lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java

```diff
@@ -1221,6 +1229,111 @@ private Location createListingLocation(String path)
+                }
+                assertTrue(fileExists(location));
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+    }
+
+    @Test
+    @Disabled("This test takes 5 minutes to run locally")
+    public void testLargeFileDoesNotExistUntilClosed()
+    {
+        assertThatNoException().isThrownBy(() -> {
+            try {
+                Location location = getRootLocation().appendPath("testFile3");
+                getFileSystem().deleteFile(location);
+                byte[] bytes = new byte[8192];
+                Slice slice = Slices.wrappedBuffer(bytes);
+                for (int i = 0; i < bytes.length / 4; i++) {
+                    slice.setInt(i * 4, i);
+                }
+                try (OutputStream outputStream = getFileSystem().newOutputFile(location).create()) {
+                    // write a 32 mb file
+                    for (int i = 0; i < 32 * 1024 * 1024 / bytes.length; i++) {
+                        outputStream.write(bytes);
+                        assertFalse(fileExists(location));
+                    }
+                }
+                assertTrue(fileExists(location));
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+    }
+
+    @Test
+    @Disabled("This test takes 10 minutes to run locally")
+    public void testLargeFileDoesNotOverwriteUntilClosed()
+    {
+        try {
+            Location location = getRootLocation().appendPath("testFile4");
+            try (OutputStream outputStream = getFileSystem().newOutputFile(location).createOrOverwrite()) {
+                outputStream.write("test".getBytes(UTF_8));
+            }
+            TrinoInputFile inputFile = getFileSystem().newInputFile(location);
+            int fileSize = toIntExact(inputFile.length());
+            byte[] fileBytes = new byte[fileSize];
+            assertEquals(fileSize, readFile(inputFile, fileBytes));
+            assertEquals("test", new String(fileBytes, UTF_8));
+            byte[] bytes = new byte[8192];
+            Slice slice = Slices.wrappedBuffer(bytes);
+            for (int i = 0; i < bytes.length / 4; i++) {
+                slice.setInt(i * 4, i);
+            }
+            try (OutputStream outputStream = getFileSystem().newOutputFile(location).createOrOverwrite()) {
+                // write a 32 mb file
+                for (int i = 0; i < 32 * 1024 * 1024 / bytes.length; i++) {
+                    outputStream.write(bytes);
+                    assertEquals(fileSize, readFile(inputFile, fileBytes));
+                    assertEquals("test", new String(fileBytes, UTF_8));
+                }
+            }
+            assertTrue(fileExists(location));
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private boolean fileExists(Location location)
```

Comment:
call this `fileExistsInListing` and add another one:
```java
private boolean fileExists(Location location)
        throws IOException
{
    return getFileSystem().newInputFile(location).exists();
}
```

in the test code assert on both